shap/plots/_style.py

```diff
@@ -0,0 +1,123 @@
+class StyleConfig:
+    """A complete set of configuration options for matplotlib-based shap plots."""
+
+    primary_color_positive: ColorType
+    primary_color_negative: ColorType
+    secondary_color_positive: ColorType
+    secondary_color_negative: ColorType
+    hlines_color: ColorType
+    vlines_color: ColorType
+    text_color: ColorType
+    tick_labels_color: ColorType
+
+
+class StyleOptions(TypedDict, total=False):
+    """A TypedDict of partial updates to a style configuration"""
+
+    # Nb. There is some duplication here with the StyleConfig dataclass, but
+    # it's necessary to provide helpful type hints with `typing.Unpack`.
+    # See https://github.com/python/typing/issues/1495
+    primary_color_positive: ColorType
+    primary_color_negative: ColorType
+    secondary_color_positive: ColorType
+    secondary_color_negative: ColorType
+    hlines_color: ColorType
+    vlines_color: ColorType
+    text_color: ColorType
+    tick_labels_color: ColorType
+
+
+_shap_defaults = StyleConfig(
+    primary_color_positive=colors.red_rgb,
+    primary_color_negative=colors.blue_rgb,
+    secondary_color_positive=colors.light_red_rgb,
+    secondary_color_negative=colors.light_blue_rgb,
+    hlines_color="#cccccc",
+    vlines_color="#bbbbbb",
+    text_color="white",
+    tick_labels_color="#999999",
+)
+
+
+def load_default_style() -> StyleConfig:
+    """Load the default style configuration."""
+    # In future, this could allow reading from a persistent config file, like matplotlib rcParams.
+    return _shap_defaults
+
+
+# Singleton instance that determines the current style.
+# CAREFUL! To ensure the correct object is picked up, do not import this directly,
+# but intead access this at runtime with get_style().
+_STYLE = load_default_style()
+
+
+def get_style() -> StyleConfig:
+    """Return all currently active global style configuration options."""
+    return _STYLE
+
+
+def set_style(**options: Unpack[StyleOptions]) -> None:
+    """Set options in the currently active global style configuration."""
+    global _STYLE
+    _STYLE = _apply_options(_STYLE, options)
+
+
+@contextmanager
+def style_context(**options: Unpack[StyleOptions]):
+    """Context manager to temporarily change style options.
+
+    NOTE: This is experimental and subject to change!
+
+    Examples
+    --------
+    To set a new style::
+
+        with shap.plots.style_context(text_color="black"):
+            shap.plots.waterfall(...)
+    """
+    global _STYLE
+    old_style = _STYLE
+    _STYLE = _apply_options(_STYLE, options)
+    yield
+    _STYLE = old_style
+
+
+def _apply_options(style: StyleConfig, changes: StyleOptions) -> StyleConfig:
+    """Return a new StyleConfig with any changes applied, handling any invalid options."""
+    valid_keys = set(f.name for f in fields(StyleConfig))
+    for key in changes.keys():
+        if key not in valid_keys:
+            raise InvalidOptionError(f"Invalid style config option: {key}")
```

Comment:
When I first saw the error I would expect that the values are checked as well. The plethora of options here is not really checkable but could we somehow just check for the correct type?

IMO we don't need to implement this now, think this is good for now

Reply:
That's a good point, I hadn't really considered it. That might be something we can come back to in a future PR to improve :)

For now, hopefully the type hints (`TypedDict` & `Unpack`) will help users validate that their changes are valid. Also, if a user _does_ provide an invalid object in a style setting, hopefully they will get an informative error message from matplotlib when they actually create a plot. But yes, type checking the inputs would be nice.

Reply:
absolutely, let's go ahead with the current implementation